### PR TITLE
[WIP] Serve KerasHub CausalLM models through vLLM on TPU

### PR DESCRIPTION
**Pull request by anthony-etim**
## Description of the change

Adds a self-contained `keras_hub/src/vllm/` module that lets any supported KerasHub `CausalLM` be served through vLLM's TPU backend (`tpu-inference`) using its native Pallas **paged-attention** kernel, reusing backbones, `from_preset()` weights, presets, and tokenizers, via a one-line API:

```python
from keras_hub.src.vllm import KerasHubLLM
llm = KerasHubLLM("keras_hub:gemma_2b_en")
print(llm.generate(["The future of AI is"])[0].outputs[0].text)
```
No core Keras layer gains a hard dependency on vLLM; the only touch points in core layers are cheap, guarded conditionals that are inert outside vLLM serving. The existing `generate()` path is unchanged.

**Why:** KerasHub generation today uses a static, dense, per-request KV cache with no continuous batching, paged attention, prefix sharing, or scheduler. This wires KerasHub into vLLM's runtime to get those, while staying on the JAX/TPU stack where Keras/JAX are the natural toolchain. The existing `generate()` path is unchanged.

**Design: contained module and guarded hooks.** No core Keras layer gains a hard dependency on vLLM; the only touch points in core layers are cheap, guarded conditionals that are inert outside vLLM serving. No hardware kernel is duplicated in KerasHub (it borrows `tpu-inference`'s RPA kernel).

New module `keras_hub/src/vllm/`:

| File | Role |
|---|---|
| `adapter.py` | `KerasVLLMAdapter`, vLLM model interface; runs the backbone so per-layer paged caches thread through; PyTorch↔JAX bridge; tied-embedding logits |
| `attention.py` | `maybe_vllm_paged_attention()`, single bridge into the RPA kernel (shape conversion, GQA inference, scale/sliding_window/soft_cap) |
| `context.py` | thread-local `vllm_context` (block tables, slot mapping, attn metadata, injected kernel fn, mesh, positions) |
| `registry.py` | `register_keras_hub_models()`, `setup_vllm_model()` (config.json and HF tokenizer export with `tokenizer.json` fallback), `KerasHubLLM` |
| `tokenizer.py` | HF tokenizer export (byte-level BPE and SentencePiece) with round-trip verification |
| `tests/` | CPU unit tests (context, bridge, tokenizer export) |

Guarded hooks in existing attention layers (no-ops outside vLLM serving): `cached_multi_head_attention.py` (GPT-2) and the `gemma`/`llama`/`mistral` attention layers gain a branch that delegates to the bridge only when a serving context is active.

**Roadmap:**
- [x] vLLM serving scaffold — thread-local serving context
- [x] Paged-attention bridge (`maybe_vllm_paged_attention`)
- [x] `KerasVLLMAdapter` (torch.nn.Module wrapping a Keras `CausalLM`)
- [x] HF tokenizer export (byte-level BPE + SentencePiece)
- [x] Registry , `setup_vllm_model` , `KerasHubLLM` one-line API
- [x] GPT-2 attention hook
- [x] Gemma / Llama / Mistral attention hooks (RoPE at vLLM positions, GQA, soft-cap)
- [x] CPU unit tests (context, bridge, tokenizer export)
- [x] Correctness: next-token-exact vs native KerasHub (bf16)
- [x] Benchmarks: pure Keras vs vLLM-native vs this integration (TTFT / throughput / latency)

**Notes / limitations:** TPU/JAX path only (`MODEL_IMPL_TYPE=vllm`); serving runs in bf16 to match the TPU paged KV cache. **Follow-up (deferred):** `gemma3`/ `gemma4` attention hooks; Gemma 2/3 soft-cap end-to-end validation; multi-chip sharding (TP>1)

## Reference

- Companion serving-side hooks: `vllm-project/tpu-inference` PR (registry recognition, Pallas-kernel injection, `soft_cap` forwarding). 
- vLLM PagedAttention / continuous batching; `tpu-inference` Ragged Paged Attention (RPA) kernel.

## Colab Notebook

**Usage:** `colab_test.ipynb` — `KerasHubLLM("keras_hub:<preset>")` end-to-end on TPU (GPT-2, Gemma 2B, Llama 3.2 1B).

**Correctness:** the integration's next token matches native KerasHub greedy generation (bf16) **exactly** for the test prompts. (Full-text greedy parity isn't the bar: bf16-on-TPU paged attention vs bf16-on-CPU dense attention diverges after the first precision-driven argmax flip — so we compare the next-token distribution in matched dtype.) CPU unit tests cover the dependency-light logic.

**Benchmarks** (`colab_benchmark.ipynb`): `gpt2_large_en` (774M) on TPU v6e, bf16, 128 output tokens, warm, best-of-3. **A** = pure KerasHub `generate()` (floor), **B** = vLLM-native (ceiling), **C** = this integration.

Throughput (output tok/s) and speedups:

| prompt (tok) | conc | A | B | C | **C/A** | **C/B** |
|---:|---:|---:|---:|---:|:---:|:---:|
| 37 | 1 | 536.9 | 359.1 | 322.7 | 0.60× | 0.90× |
| 37 | 32 | 2329 | 5642 | 5517 | **2.37×** | 0.98× |
| 37 | 64 | 2475 | 6175 | 6174 | **2.49×** | 1.00× |
| 37 | 128 | 4751 | 6589 | 6559 | 1.38× | 1.00× |
| 602 | 1 | 442.1 | 330.4 | 327.8 | 0.74× | 0.99× |
| 602 | 32 | 583.0 | 5329 | 5246 | **9.00×** | 0.98× |
| 602 | 64 | **OOM** | 5923 | 5888 | **A-OOM** | 0.99× |
| 602 | 128 | **OOM** | 6341 | 6294 | **A-OOM** | 0.99× |

Time-to-first-token (p50, s):

| prompt (tok) | conc | A | B | C |
|---:|---:|---:|---:|---:|
| 37 | 1 | 0.0045 | 0.0095 | 0.012 |
| 37 | 128 | 0.041 | 0.134 | 0.130 |
| 602 | 1 | 0.0085 | 0.011 | 0.011 |
| 602 | 32 | 0.150 | 0.062 | 0.064 |
| 602 | 64 | OOM | 0.113 | 0.118 |
| 602 | 128 | OOM | 0.225 | 0.229 |

**Takeaways:** C ≈ B on every metric (C/B ≈ 0.98–1.00× past c=1) → the adapter adds no measurable overhead vs native vLLM. C beats pure Keras under load (2.4× at moderate concurrency, **9× at long-context c=32**, serves loads where pure Keras **OOMs**; TTFT ~2.4× lower at 602/c=32). Pure Keras wins only at c=1 (single-request) — the regime paged serving isn't built to optimize.

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

**Comment by anthony-etim** [comment]
Superseding with a native flax/nnx implementation (no torch/torchax)
A new PR will follow shortly.
